internal/cmd/test.go

```diff
@@ -167,23 +167,32 @@ func RunCommonCMDTests(test *testing.T, mainFunc func(), args []string, commonTe
 	stdout, stderr, exitCode, err := RunCMDTest(test, mainFunc, args, commonTestCase.LogLevel)
 	if err != nil {
 		test.Errorf("%sCMD run returned an error: '%v'.", prefix, err)
+		logOutputs(test, stdout, stderr)
 		return "", "", -1, false
 	}
 
 	if commonTestCase.ExpectedExitCode != exitCode {
 		test.Errorf("%sUnexpected exit code. Expected: '%d', Actual: '%d'.", prefix, commonTestCase.ExpectedExitCode, exitCode)
+		logOutputs(test, stdout, stderr)
 		return "", "", -1, false
 	}
 
 	if !strings.Contains(stderr, commonTestCase.ExpectedStderrSubstring) {
 		test.Errorf("%sUnexpected stderr substring. Expected stderr substring: '%s', Actual stderr: '%s'.", prefix, commonTestCase.ExpectedStderrSubstring, stderr)
+		logOutputs(test, stdout, stderr)
 		return "", "", -1, false
 	}
 
 	if !commonTestCase.IgnoreStdout && commonTestCase.ExpectedStdout != stdout {
 		test.Errorf("%sUnexpected output. Expected: \n'%s', \n Actual: \n'%s'.", prefix, commonTestCase.ExpectedStdout, stdout)
+		logOutputs(test, stdout, stderr)
 		return "", "", -1, false
 	}
 
 	return stdout, stderr, exitCode, true
 }
+
+func logOutputs(test *testing.T, stdout string, stderr string) {
+	test.Logf("\n--- START OF STDOUT ---\n%s\n--- END OF STDOUT---", stdout)
+	test.Logf("\n--- START OF STDERR ---\n%s\n--- END OF STDERR ----", stderr)
```

Comment:
I'm not a fan of the formatting here:
 - Starting with a newline is a bit strange.
 - The "start of"/"end of" wording seems a bit unnecessary (this is the output of a failing test, not something sent to a normal user).
 - Missing space after "END OF STDOUT".

We already have an example for this stdout/stderr pattern in a test:
https://github.com/edulinq/autograder-server/blob/main/internal/grader/grade_test.go#L82